ARCs/arc-0060.md

```diff
@@ -0,0 +1,330 @@
+---
+arc: 60
+title: Algorand Wallet Structured Data Signing API
+description: API function fot signing structured data
+author: Stefano De Angelis (@deanstef)
+discussions-to: https://github.com/algorandfoundation/ARCs/issues/284
+status: Draft
+type: Standards Track
+category: Interface
+created: 2024-02-28
+requires: 1
+---
+
+# Algorand Wallet Structured Data Signing API
+
+> This ARC is inspired by [ARC-1](./arc-0001.md).
+
+## Abstract
+
+ARC-1 defines a standard for signing Algorand transactions that are represented as structured objects. This proposal extends the signing to generic byte arrays encoded with a standardized structure.
+
+[ARC-60](./arc-0060.md) defines an API for wallets to sign structured data that are not Algorand transactions.
+
+## Motivation
+
+Signing data is a common and critical operation. Users may need to sign data for multiple reasons (e.g. delegate signatures, DIDs, signing documents, authentication).
+
+Algorand wallets need a standard approach to byte signing to unlock self-custodial services and protect users from malicious and attack-prone signing workflows.
+
+This ARC provides a standard API for bytes signing. The API encodes byte arrays into well structured JSON schemas together with additional metadata. It requires wallets to validate the signing inputs, notify users about what they are signing and warn them in case of dangerous signing requests.
+
+## Specification
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+> Comments like this are non-normative
+
+### Overview
+
+> This section is non-normative
+
+Signatures of bytes are processed with the `signData(data, metadata, signer)` function.
```

Comment:
@deanstef requiring the public key (a.k.a signer) would be somewhat cumbersome for HDWallets to find the correct address to use. Would have to do a recursive search. 

Legacy algorand wallets are single address, with HDWallets they will support BIP44 and providing the path  as optional would suit best.

Thoughts?

Reply:
In the first version of ARC-60 there was the possibility of passing a HDWallet derivation path as signer parameter (see https://github.com/deanstef/ARCs/blob/67754a89ff9e211cfee6f29ebd7aa4f016b19243/ARCs/arc-0060.md?plain=1#L122)

Then we decided to keep it simple and just opt for passing the public key of the signer. I'd be happy to re-introduce HDWallets support though

Reply:
I understand. Although just the public key is ambiguous when you have N possible key pairs in a BIP32 / BIP44 path scheme.

Reply:
This has been addressed